File: medium/Jump_Game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if nums.count(0) == 0 or len(nums) == 1:
            return True
        target_index = 0
        while target_index < len(nums):
            while nums[target_index] != 0 and target_index < len(nums)-1:
                target_index+=1
            k = target_index
            for i, num in enumerate(nums[:target_index][::-1]):
                i+=1
                if num > i or num == i and target_index == len(nums)-1:
                    break
                else:
                    k-=1
            if k == 0:
                return False
            target_index+=1
        return True

File: cpp/medium/Jump_Game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        if (nums[0] == 0 && nums.size() != 1) return false;
        int max_idx = 1;
        int idx = 0;
        while (idx <= max_idx) {
            max_idx = max(idx+nums[idx], max_idx);
            if (max_idx >= nums.size()-1) return true;
            idx++;
        }
        return false;
    }
};
